DirectProgramming/DPC++FPGA/Tutorials/Features/private_copies/src/private_copies.cpp

```diff
@@ -0,0 +1,211 @@
+// This FPGA best practice reduces name mangling in the optimization reports.
+template <int NumCopies>
+class Kernel;
+
+// Helper function to compute some simple math operations
+int SimpleMath(int input_data1, int input_data2) {
+  int temp_result = int(sqrt(input_data1)) + int(log(input_data2));
+  return int(cos(temp_result)) + int(sin(input_data2)) + int(sqrt(input_data1));
+}
+
+// Launch a kernel on the device specified by selector.
+// The kernel's functionality is designed to show the
+// performance impact of the private_copies attribute.
+template <int NumCopies>
+void SimpleMathWithShift(const device_selector &selector, const IntArray &array,
+                         int shift, IntScalar &result) {
+  double kernel_time = 0.0;
+
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer buffer_array(array);
+    buffer<int, 1> buffer_result(result.data(), 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessor_array(buffer_array, h, read_only);
+      accessor accessor_result(buffer_result, h, write_only, noinit);
+
+      h.single_task<Kernel<NumCopies>>([=]() [[intel::kernel_args_restrict]] {
+        int r = 0;
+
+        for (size_t i = 0; i < kMaxIter; i++) {
+          [[intel::private_copies(NumCopies)]] int a[kSize];
+          [[intel::private_copies(NumCopies)]] int b[kSize];
+          int c[kSize];
+          for (size_t j = 0; j < kSize; j++) {
+            a[j] = accessor_array[(i * 4 + j) % kSize] * shift;
+          }
+          for (size_t j = 0; j < kSize; j++)
+            b[j] = SimpleMath(a[j], a[kSize - j - 1]);
+          for (size_t j = 0; j < kSize; j++)
+            c[j] = SimpleMath(b[j], b[kSize - j - 1]);
+          for (size_t j = 0; j < kSize; j++) 
+            r += c[j];
+        }
+
+        accessor_result[0] = r;
+      });
+    });
+
+    // SYCL event profiling allows the kernel execution to be timed
+    double start = e.get_profiling_info<info::event_profiling::command_start>();
+    double end = e.get_profiling_info<info::event_profiling::command_end>();
+    kernel_time = (double)(end - start) * 1e-6;
+
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+
+  // The performance of the kernel is measured in GFlops, based on:
+  // 1) the number of inting-point operations performed by the kernel.

```

Comment:
I've never seen inting-point before ! 
A quick google search does not return anything 
I would suggest changing it to integer operations :)